internal/configs/provider_validation.go

```diff
@@ -831,3 +842,165 @@ func providerName(name, alias string) string {
 	}
 	return name
 }
+
+func providerIterationIdenticalWarning(blockType string, sourceExpr, instanceExpr hcl.Expression) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+	if providerIterationIdentical(sourceExpr, instanceExpr) {
+		// foot, meet gun
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagWarning,
+			Summary:  "Likely misconfiguration of provider iteration",
+			Detail:   fmt.Sprintf("Provider and %s both share identical iteration expressions, this is not recommended. When a key is removed, the corresponding resources will no longer have a provider that is able to destroy them.\nInstead, it is recommended to have %s's iteration be a subset of the provider's iteration expression. See the OpenTofu documentation for more details.\n\nTo disable this warning, wrap one of the expressions in a function like coalesce() to disable the check.", blockType, blockType),
+			Subject:  sourceExpr.Range().Ptr(),
+		})
+	}
+	return diags
+}
+
+// Might have gone a bit overboard on this...
+//
+//nolint:funlen,gocognit,gocyclo,cyclop // just a lot of branches
+func providerIterationIdentical(a, b hcl.Expression) bool {
+	if a == nil && b == nil {
+		return true
+	}
+	// Remove parenthesis
+	if ae, ok := a.(*hclsyntax.ParenthesesExpr); ok {
+		return providerIterationIdentical(ae.Expression, b)
+	}
+	if be, ok := b.(*hclsyntax.ParenthesesExpr); ok {
+		return providerIterationIdentical(a, be.Expression)
+	}
+
+	if ae, ok := a.(*hclsyntax.ObjectConsKeyExpr); ok {
+		return providerIterationIdentical(ae.Wrapped, b)
+	}
+	if be, ok := b.(*hclsyntax.ObjectConsKeyExpr); ok {
+		return providerIterationIdentical(a, be.Wrapped)
+	}
+
+	switch as := a.(type) {
+	case *hclsyntax.ScopeTraversalExpr:
+		if bs, bok := b.(*hclsyntax.ScopeTraversalExpr); bok {
+			return TraversalStr(as.Traversal) == TraversalStr(bs.Traversal)
```

Comment:
It seems that right now we don't have a consistent idea of what exactly we're trying to catch vs. not trying to catch. I'd personally rather have a simple rule that we can write once and never maintain again vs. something that is so in the weeds of HCL that it's likely to need changing every time HCL changes. If we can't get what we need without coming close to a full HCL tree traversal then I'd hope we could find a different way to solve this problem that is less "magical".

Since we don't seem to be aligned on what exactly our goal is here, perhaps we should switch to discussing an addition to the RFC first so we can nail down what we're intending to solve for vs. not intending to solve for. Even if we end up just writing down the same thing you already implemented, it will at least put in writing what contract we're intending to provide here so that future maintainers can know what they are expected to retain support for. :thinking: